src/shell/commands/disk_rebalance.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#include "shell/commands.h"
+#include "shell/argh.h"
+
+#include <math.h>
+#include <fmt/ostream.h>
+#include <dsn/utility/errors.h>
+#include <dsn/utility/output_utils.h>
+#include <dsn/utility/string_conv.h>
+#include <dsn/dist/replication/duplication_common.h>
+
+bool query_disk_info(command_executor *e, shell_context *sc, arguments args)
+{
+    // disk_info [-n|--node node_address] [-a|--app app_replica_count]
+
+    argh::parser cmd(args.argc, args.argv);
+    if (cmd.pos_args().size() > 3) {
+        fmt::print(stderr, "too many params\n");
+        return false;
+    }
+
+    bool query_one_node = cmd[{"-n", "--node"}];
+    bool query_app_replica_count = cmd[{"-a", "--app_replica"}];
+
+    std::map<dsn::rpc_address, dsn::replication::node_status::type> nodes;
+    auto error = sc->ddl_client->list_nodes(::dsn::replication::node_status::NS_INVALID, nodes);
+    if (error != dsn::ERR_OK) {
+        std::cout << "list nodes failed, error=" << error.to_string() << std::endl;
+        return false;
+    }
+
+    std::vector<dsn::rpc_address> targets;
+    if (query_one_node) {
+        if (!cmd(1)) {
+            fmt::print(stderr, "missing param <node_address>\n");
+            return false;
+        }
+        std::string node_address = cmd(1).str();
+
+        for (auto &node : nodes) {
+            if (node.first.to_std_string() == node_address) {
+                targets.emplace_back(node.first);
+            }
+        }
+
+        if (targets.empty()) {
+            fmt::print(stderr, "please input valid target node_address!\n");
+            return false;
+        }
+    } else {
+        if (query_app_replica_count) {
+            fmt::print(stderr, "please input node_address when query app replica count!\n");
+            return false;
+        }
+        for (const auto &node : nodes) {
+            targets.emplace_back(node.first);
+        }
+    }
+
+    int app_id = 0;
+    if (query_app_replica_count) {
+        if (cmd(2)) {
+            app_id = stoi(cmd(2).str());
+        }
+    }
+
+    const auto &err_resps = sc->ddl_client->query_disk_info(targets, app_id);
+
+    dsn::utils::table_printer node_printer;
+    node_printer.add_title("node");
+    node_printer.add_column("total_capacity(MB)");
+    node_printer.add_column("avalable_capacity(MB)");
+    node_printer.add_column("avalable_ratio(%)");
+    node_printer.add_column("capacity_balance");
+
+    int total_capacity_ratio = 0;
```

Comment:
a node, the disk info focuses node, but not cluster. `node_printer` also illustrate it.

Reply:
So why not define it in the loop?

Reply:
The follow code:
```c++
for (const auto &disk_info : resp.disk_infos) {
      int disk_available_ratio = disk_info.disk_capacity_mb == 0
                                      ? 0
                                      : std::round(disk_info.disk_available_mb *
                                      100.0 / disk_info.disk_capacity_mb);
      int disk_density = disk_available_ratio - total_capacity_ratio;
...
```
 will user it, Otherwise it need to be calc two time.